docs/architecture/adr-006-support-multisig-chain-link.md

```diff
@@ -0,0 +1,320 @@
+				return err
+			}
+
+			cdc, _ := app.MakeCodecs()
+			bz, err := cdc.MarshalJSON(&chainLinkJSON)
+			if err != nil {
+				return err
+			}
+
+			if filename != "" {
+				err = ioutil.WriteFile(filename, bz, 0600)
+				if err != nil {
+					return err
+				}
+			}
+			cmd.Println(fmt.Sprintf("Chain link JSON file stored at %s", filename))
+			return nil
+		},
+	}
+	cmd.Flags().String(flags.FlagChainID, "", "network chain ID")
+	return cmd
+} 
+
+// createSingleSigAccountChainLinkJSON creates the chain-link JSON for single signature account via getter
+func createSingleSigChainLinkJSON(
+	getter chainlinktypes.ChainLinkReferenceGetter
+) (profilescliutils.ChainLinkJSON, error) {
+	mnemonic, err := getter.GetMnemonic()
+	if err != nil {
+		return err
+	}
+
+	chain, err := getter.GetChain()
+	if err != nil {
+		return err
+	}
+
+	return generateChainLinkJSON(mnemonic, chain)
+}
+
+// createMultiSigChainLinkJSON creates the chain-link JSON for multi signature account via getter
+func createMultiSigChainLinkJSON(
+	cmd *cobra.Command, getter chainlinktypes.ChainLinkReferenceGetter
+)(profilescliutils.ChainLinkJSON, error) {
+
+	txFile, err := getter.GetTxFile()
+	if err != nil {
+		return err
+	}
+
+	multisignFile, err := getter.GetMultiSignFile()
+	if err != nil {
+		return err
+	}
+
+	chainID, err := getter.GetChainID()
+	if err != nil {
+		return err
+	}
+
+	return getChainLinkJSONFromMultiSign(
+		cmd,
+		txFile,
+		multiSign, 
+		chain,
+	)
+}
+```
+
+We propose to create chain-link JSON from the multisign file. In cosmos-sdk, multisig account transaction signing depends on `tx sign` and 
+`tx multisign` commands. To send a transaction to the node, the multisig account owners should create a raw transaction file, then the threshold 
+number of them sign it to generate the signed files by using the `tx sign` command with their keys. Subsequently, one of them gathers all the signed 
+files and uses the `tx multisign` command with the raw transaction file to get the multisign file. The multisign file includes not only all public keys 
+and threshold of the multisig account but also the required number of signatures so that creating the chain-link JSON from the multisign file is possible.
+
+The whole process in code is presented below:
+```go
+// getChainLinkJSONFromMultiSign generates the chain-link JSON from the multisign file and its raw transaction file
+func getChainLinkJSONFromMultiSign(
+	cmd *cobra.Command,
+	txFile string,
+	multiSignFile string,
+	chainID string,
+	chain chainlinktypes.Chain,
+) (profilescliutils.ChainLinkJSON, error) {
+	clientCtx, err := client.GetClientTxContext(cmd)
+	if err != nil {
+		return profilescliutils.ChainLinkJSON{}, err
+	}
+
+	parsedTx, err := authclient.ReadTxFromFile(clientCtx, txFile)
+	if err != nil {
+		return profilescliutils.ChainLinkJSON{}, err
+	}
+
+	parsedMultiTx, err := authclient.ReadTxFromFile(clientCtx, multiSignFile)
+	if err != nil {
+		return profilescliutils.ChainLinkJSON{}, err
+	}
```

Comment:
Why is it necessary to have two different files? Can't we simply use the signed transaction file?

Reply:
Because the raw tx file is the plaintext of the multisigned file.

Reply:
@RiccardoM Because the raw tx file is the plaintext of the multisigned file.

Reply:
@dadamu Can we have an example of contents for both files? I would like to better understand this

Reply:
@RiccardoM To generate a multisign file, the process is as the follows:
1. One of signers creates the unsigned tx file as plain text, the unsigned  tx file is like:
```JSON
{
    "body": {
        "messages": [
            {
                "@type": "/cosmos.gov.v1beta1.MsgVote",
                "proposal_id": "1",
                "voter": "desmos140r08h3y5he52eg7xckf2uuldweekjz536lada",
                "option": "VOTE_OPTION_YES"
            }
        ],
        "memo": "",
        "timeout_height": "0",
        "extension_options": [],
        "non_critical_extension_options": []
    },
    "auth_info": {
        "signer_infos": [],
        "fee": {
            "amount": [],
            "gas_limit": "200000",
            "payer": "",
            "granter": ""
        }
    },
    "signatures": []
}
```

2. The threshold number of signers sign the unsigned tx file to generate signed tx files, the signed tx file is as the follows:
```JSON
{
    "signatures": [
        {
            "public_key": {
                "@type": "/cosmos.crypto.secp256k1.PubKey",
                "key": "AhpowBmBeBIdMnkO0Mu0MgkrL3j9PRkqRu7INXnUx6/+"
            },
            "data": {
                "single": {
                    "mode": "SIGN_MODE_LEGACY_AMINO_JSON",
                    "signature": "iejF2C2/ITAjGqJkw90Zt0aFVzOzpnmi6cETKqoFJHg+6GMhdW8xFEXthfSa9AbdyjAJZN1P9PsO8srL6qp4eg=="
                }
            },
            "sequence": "0"
        }
    ]
}
```

3.  Use multisign command gathered all signatures of signed tx files then building multisigned_file from the unsigned tx file, the multisigned_file is as below:
```JSON
{
    "body": {
        "messages": [
            {
                "@type": "/cosmos.gov.v1beta1.MsgVote",
                "proposal_id": "1",
                "voter": "desmos140r08h3y5he52eg7xckf2uuldweekjz536lada",
                "option": "VOTE_OPTION_YES"
            }
        ],
        "memo": "",
        "timeout_height": "0",
        "extension_options": [],
        "non_critical_extension_options": []
    },
    "auth_info": {
        "signer_infos": [
            {
                "public_key": {
                    "@type": "/cosmos.crypto.multisig.LegacyAminoPubKey",
                    "threshold": 2,
                    "public_keys": [
                        {
                            "@type": "/cosmos.crypto.secp256k1.PubKey",
                            "key": "AhpowBmBeBIdMnkO0Mu0MgkrL3j9PRkqRu7INXnUx6/+"
                        },
                        {
                            "@type": "/cosmos.crypto.secp256k1.PubKey",
                            "key": "Ax4ee7c3mJHeiv50SEDfbUe7GpT+li+BPepH3rBUP6it"
                        },
                        {
                            "@type": "/cosmos.crypto.secp256k1.PubKey",
                            "key": "A7juoEs01zEPSX1JmZvMz9HNOOiQqKLsfak7oygTztOD"
                        }
                    ]
                },
                "mode_info": {
                    "multi": {
                        "bitarray": {
                            "extra_bits_stored": 3,
                            "elems": "gA=="
                        },
                        "mode_infos": [
                            {
                                "single": {
                                    "mode": "SIGN_MODE_LEGACY_AMINO_JSON"
                                }
                            }
                        ]
                    }
                },
                "sequence": "0"
            }
        ],
        "fee": {
            "amount": [],
            "gas_limit": "200000",
            "payer": "",
            "granter": ""
        }
    },
    "signatures": [
        "CkCJ6MXYLb8hMCMaomTD3Rm3RoVXM7OmeaLpwRMqqgUkeD7oYyF1bzEURe2F9Jr0Bt3KMAlk3U/0+w7yysvqqnh6"
    ]
}
```

References: 
[multisign command](https://github.com/cosmos/cosmos-sdk/blob/99e2e0f060b6675a449cd07147ca0bc96c577e56/x/auth/client/cli/tx_multisign.go#L72)
[tx sign command](https://github.com/cosmos/cosmos-sdk/blob/2b02f066aae39cd2ece720e2616acf5b3308bf8e/x/auth/client/cli/tx_sign.go#L209)

Reply:
@RiccardoM Yeah, it might be possible to use only one file. Let me try a thing.

Reply:
@RiccardoM Confirmed, we can use multisigned_tx file only to create the proof.